lib/internal/fs/utils.js

```diff
@@ -31,24 +33,139 @@ const {
   S_IFREG,
   S_IFSOCK,
   UV_FS_SYMLINK_DIR,
-  UV_FS_SYMLINK_JUNCTION
+  UV_FS_SYMLINK_JUNCTION,
+  UV_DIRENT_UNKNOWN,
+  UV_DIRENT_FILE,
+  UV_DIRENT_DIR,
+  UV_DIRENT_LINK,
+  UV_DIRENT_FIFO,
+  UV_DIRENT_SOCKET,
+  UV_DIRENT_CHAR,
+  UV_DIRENT_BLOCK
 } = process.binding('constants').fs;
 
 const isWindows = process.platform === 'win32';
 
+let fs;
+function lazyLoadFs() {
+  if (!fs) {
+    fs = require('fs');
+  }
+  return fs;
+}
+
 function assertEncoding(encoding) {
   if (encoding && !Buffer.isEncoding(encoding)) {
     throw new ERR_INVALID_OPT_VALUE_ENCODING(encoding);
   }
 }
 
+class Dirent {
+  constructor(name, type) {
+    this.name = name;
+    this[kType] = type;
+  }
+
+  isDirectory() {
+    return this[kType] === UV_DIRENT_DIR;
+  }
+
+  isFile() {
+    return this[kType] === UV_DIRENT_FILE;
+  }
+
+  isBlockDevice() {
+    return this[kType] === UV_DIRENT_BLOCK;
+  }
+
+  isCharacterDevice() {
+    return this[kType] === UV_DIRENT_CHAR;
+  }
+
+  isSymbolicLink() {
+    return this[kType] === UV_DIRENT_LINK;
+  }
+
+  isFIFO() {
+    return this[kType] === UV_DIRENT_FIFO;
+  }
+
+  isSocket() {
+    return this[kType] === UV_DIRENT_SOCKET;
+  }
+
+  isUnknown() {
+    return this[kType] === UV_DIRENT_UNKNOWN;
+  }
+}
+
+class DirentFromStats extends Dirent {
+  constructor(name, stats) {
+    super(name, null);
+    this[kStats] = stats;
+  }
+}
+
+for (const name of Reflect.ownKeys(Dirent.prototype)) {
+  if (name === 'isUnknown' || name === 'constructor') {
+    continue;
+  }
+  DirentFromStats.prototype[name] = function() {
+    return this[kStats][name]();
+  };
+}
+
 function copyObject(source) {
   var target = {};
   for (var key in source)
     target[key] = source[key];
   return target;
 }
 
+function getDirents(path, [names, types], callback) {
+  var i;
+  if (typeof callback == 'function') {
+    const len = names.length;
+    let toFinish = 0;
+    for (i = 0; i < len; i++) {
+      const type = types[i];
+      if (type === UV_DIRENT_UNKNOWN) {
+        const name = names[i];
+        const idx = i;
+        toFinish++;
+        lazyLoadFs().stat(pathModule.resolve(path, name), (err, stats) => {
```

Comment:
Are we sure we want to have this behavior?

At a minimum, for me, DirEntry is associated with entry information, including information that there is no type information for a given entry.

It's confusing.

Reply:
This behavior was specifically asked for by @addaleax in this comment: https://github.com/nodejs/node/pull/22020#discussion_r207039181

`isUnknown()` will still return true for entries whose type information was retrieved via `fs.stat`, so for the use case you're describing, information is not lost.

Reply:
So, in this case, the `isUnknown` method is not described in the documentation (`doc/api/fs.md` `Dirent` class).

But, unfortunately, I still don't understand why the low-level tool has an implicit action.

For example, I intended to make additional logic for unknown-entries in the package:

> Read the directory. Get a set of entries. If the user wants to see `fs.Stats` for entries, then we get it.

In the current situation we will receive the `fs.Stats` twice (no big deal) if the system doesn't give us types to entries. I understand that this is a degenerate example.

P.S.: That's interesting, can the same system provide a type for some entries, and for some not.

❤️ In any case, thank you for your work!

Reply:
> So, in this case, the `isUnknown` method is not described in the documentation (`doc/api/fs.md` `Dirent` class).

That's an mistake on my part. Whoops! I'll add it in.

> But, unfortunately, I still don't understand why the low-level tool has an implicit action.

The intent (I think, based on the comment by @addaleax) is to actually have directory entry type information in as many cases as possible, and not depend on the user code to perform an extra action to get the information they asked for in the first place in the event of an easily recoverable bad case (i.e. `isUnknown`). This seems reasonable to me.

> In the current situation we will receive the fs.Stats twice (no big deal) if the system doesn't give us types to entries. I understand that this is a degenerate example.

Perhaps a future optimization PR could cache the results of the first `stat` call? (Maybe clearing that cache on the next tick?)

> P.S.: That's interesting, can the same system provide a type for some entries, and for some not.

Excellent question! To be honest, I'm not really sure what the exhaustive list of situations is, in which `isUnknown` returns true. I do know that it seemed to show up reliably in CI on SmartOS and AIX for entries that I knew were ordinary files, and `fs.stat` works just fine on them. I'm basically just going with what libuv reports.

Reply:
@mrmlnc Sorry I looked over my code again and it seems that my intention was actually to *remove* the `isUnknown` method, because it's now unnecessary, since the underlying `fs.stat` call will make the type always knowable. I think that's the best path to take here, so I'll go ahead and remove `isUnknown` from the implementation and the test file.

That being said, if you do find yourself in a situation where you want to know whether `fs.stat` has been called internally, note that the `Dirent` instances in that case will actually be instances of `DirentFromStats`, an undocumented subclass I used in order to make it easy to have the methods from `Stats` fill in the missing data. So you can always detect that if you really need that information.

Sorry for the confusion! It's getting late here (UTC-7), haha.